File: tools/garmin_calendar/garmin_workout_builder.py
def timed_step(order, step_type, seconds):
    return {
        "type": "ExecutableStepDTO",
        "stepOrder": order,
        "stepType": {
            "stepTypeId": step_type,
        },
        "endCondition": {
            "conditionTypeId": 2,
            "conditionTypeKey": "time",
        },
        "endConditionValue": seconds,
        "targetType": {
            "workoutTargetTypeId": 1,
            "workoutTargetTypeKey": "no.target",
        },
    }


def hr_step(order, step_type, seconds, low=None, high=None):
    step = timed_step(order, step_type, seconds)

    if low is not None or high is not None:
        step["targetType"] = {
            "workoutTargetTypeId": 4,
            "workoutTargetTypeKey": "heart.rate.zone",
        }

        if low is not None:
            step["targetValueOne"] = low

        if high is not None:
            step["targetValueTwo"] = high

    return step




def sport_type_for_modality(modality):
    normalized = modality.lower()

    if "run" in normalized:
        return {
            "sportTypeId": 1,
            "sportTypeKey": "running",
        }

    if "c2 bike" in normalized or normalized == "bike":
        return {
            "sportTypeId": 2,
            "sportTypeKey": "cycling",
        }

    return {
        "sportTypeId": 6,
        "sportTypeKey": "cardio_training",
    }

def build_matt_row_workout(workout_date):
    return {
        "workoutName": "M.A.T.T. Row Test",
        "estimatedDurationInSecs": 2400,
        "sportType": {
            "sportTypeId": 6,
            "sportTypeKey": "fitness_equipment",
        },
        "workoutSegments": [
            {
                "segmentOrder": 1,
                "sportType": {
                    "sportTypeId": 6,
                    "sportTypeKey": "fitness_equipment",
                },
                "workoutSteps": [
                    timed_step(1, 3, 600.0),
                    timed_step(2, 3, 1200.0),
                    timed_step(3, 3, 600.0),
                ],
            }
        ],
    }


def build_z2_steps(candidate, age):
    steps = []
    order = 1

    hr_160 = 160 - age
    hr_165 = 165 - age
    hr_170 = 170 - age
    hr_180 = 180 - age

    # Warm up: build progressively toward 160/165/170 - age.
    steps.append(hr_step(order, 1, 300, low=100, high=hr_160))
    order += 1

    steps.append(hr_step(order, 1, 300, low=hr_160, high=hr_165))
    order += 1

    steps.append(hr_step(order, 1, 300, low=hr_165, high=hr_170))
    order += 1

    # Working window: 170 - age through 180 - age.
    steps.append(
        hr_step(
            order,
            3,
            candidate["work_seconds"],
            low=hr_170,
            high=hr_180,
        )
    )
    order += 1

    # Cool down: reverse the warm-up progression.
    steps.append(hr_step(order, 2, 300, low=hr_165, high=hr_170))
    order += 1

    steps.append(hr_step(order, 2, 300, low=hr_160, high=hr_165))
    order += 1

    steps.append(hr_step(order, 2, 300, low=100, high=hr_160))

    return steps



def build_z2_workout(candidate, age):
    steps = build_z2_steps(candidate, age)

    modality = candidate["modality"].lower()

    if "run" in modality:
        sport_type = {
            "sportTypeId": 1,
            "sportTypeKey": "running",
        }
    elif "bike" in modality:
        sport_type = {
            "sportTypeId": 2,
            "sportTypeKey": "cycling",
        }
    else:
        sport_type = {
            "sportTypeId": 6,
            "sportTypeKey": "cardio_training",
        }

    return {
        "workoutName": f"Z2 {candidate['modality']} - {candidate['workout_date']}",
        "sportType": sport_type,
        "workoutSegments": [
            {
                "segmentOrder": 1,
                "sportType": sport_type,
                "workoutSteps": steps,
            }
        ],
    }



def pace_to_speed(pace, distance_meters):
    parts = pace.strip().split(":")

    if len(parts) != 2:
        raise ValueError(f"Invalid pace: {pace}")

    minutes = int(parts[0])
    seconds = float(parts[1])
    total_seconds = minutes * 60 + seconds

    if total_seconds <= 0:
        raise ValueError(f"Invalid pace: {pace}")

    return distance_meters / total_seconds


def pace_to_meters_per_second(pace):
    return pace_to_speed(pace, 1609.344)


def apply_run_pace_target(step, low_pace, high_pace):
    speeds = [
        pace_to_meters_per_second(low_pace),
        pace_to_meters_per_second(high_pace),
    ]

    step["targetType"] = {
        "workoutTargetTypeId": 6,
        "workoutTargetTypeKey": "pace.zone",
    }
    step["targetValueOne"] = max(speeds)
    step["targetValueTwo"] = min(speeds)
    return step


def format_gear_target(target):
    low = target["low"]
    high = target["high"]
    value = low if low == high else f"{low}-{high}"

    units = {
        "minPerMile": "min/mile",
        "minPer500m": "min/500m",
        "minPer1000m": "min/1000m",
        "rpm": "RPM",
    }

    return f"Target {value} {units[target['metric']]}"


def modality_uses_structured_target(modality):
    normalized = modality.lower()

    return (
        "run" in normalized
        or "row" in normalized
        or "bike" in normalized
    ) and "echo" not in normalized


def apply_gear_target(
    step,
    target,
    enforce=True,
):
    metric = target["metric"]
    low = target["low"]
    high = target["high"]

    step["description"] = format_gear_target(target)

    if not enforce:
        return step

    if metric == "minPerMile":
        apply_run_pace_target(step, low, high)

    elif metric in {"minPer500m", "minPer1000m"}:
        distance = 500 if metric == "minPer500m" else 1000
        speeds = [
            pace_to_speed(low, distance),
            pace_to_speed(high, distance),
        ]

        step["targetType"] = {
            "workoutTargetTypeId": 5,
            "workoutTargetTypeKey": "speed.zone",
        }
        step["targetValueOne"] = min(speeds)
        step["targetValueTwo"] = max(speeds)

    else:
        raise ValueError(
            f"Structured target is not supported for metric: {metric}"
        )

    return step


def build_gear_steps(candidate):
    steps = []
    order = 1

    for round_number in range(candidate["rounds"]):
        work_step = timed_step(
            order,
            3,
            candidate["work_seconds"],
        )

        target = candidate.get("gear_target")

        if target:
            apply_gear_target(
                work_step,
                target,
                enforce=modality_uses_structured_target(
                    candidate["modality"]
                ),
            )

        elif (
            candidate["modality"].lower() == "run"
            and candidate.get("pace_low")
            and candidate.get("pace_high")
        ):
            apply_run_pace_target(
                work_step,
                candidate["pace_low"],
                candidate["pace_high"],
            )

        steps.append(work_step)
        order += 1

        # No recovery after the final work interval.
        if round_number < candidate["rounds"] - 1:
            steps.append(
                timed_step(order, 4, candidate["rest_seconds"])
            )
            order += 1

    return steps


def build_gear_workout(candidate):
    steps = build_gear_steps(candidate)

    sport_type = sport_type_for_modality(
        candidate["modality"]
    )

    return {
        "workoutName": (
            f"{candidate['prescription']} {candidate['modality']} - {candidate['workout_date']}"
        ),
        "sportType": sport_type,
        "estimatedDurationInSecs": sum(
            step["endConditionValue"] for step in steps
        ),
        "workoutSegments": [
            {
                "segmentOrder": 1,
                "sportType": sport_type,
                "workoutSteps": steps,
            }
        ],
    }


def build_mixed_gear_workout(candidate):
    workout_steps = []

    for order, candidate_step in enumerate(
        candidate["steps"],
        start=1,
    ):
        is_work = candidate_step["kind"] == "work"
        step = timed_step(
            order,
            3 if is_work else 4,
            candidate_step["seconds"],
        )

        modality = candidate_step.get("modality") or "Recovery"
        target = candidate_step.get("gear_target")

        if is_work and target:
            apply_gear_target(
                step,
                target,
                enforce=modality_uses_structured_target(modality),
            )
            step["description"] = (
                f"{modality} - {format_gear_target(target)}"
            )
        else:
            step["description"] = modality

        workout_steps.append(step)

    sport_type = {
        "sportTypeId": 6,
        "sportTypeKey": "cardio_training",
    }

    return {
        "workoutName": (
            f"{candidate['prescription']} "
            f"{candidate['modality']} - "
            f"{candidate['workout_date']}"
        ),
        "sportType": sport_type,
        "estimatedDurationInSecs": sum(
            step["endConditionValue"]
            for step in workout_steps
        ),
        "workoutSegments": [
            {
                "segmentOrder": 1,
                "sportType": sport_type,
                "workoutSteps": workout_steps,
            }
        ],
    }


def build_power_steps(candidate):
    steps = []
    order = 1

    for round_number in range(candidate["rounds"]):
        steps.append(
            timed_step(order, 3, candidate["work_seconds"])
        )
        order += 1

        # Recovery is the remainder of the "Every X:XX" interval.
        # No recovery is needed after the final work interval.
        if (
            round_number < candidate["rounds"] - 1
            and candidate["recovery_seconds"] > 0
        ):
            steps.append(
                timed_step(order, 4, candidate["recovery_seconds"])
            )
            order += 1

    return steps


def build_power_workout(candidate):
    steps = build_power_steps(candidate)

    sport_type = sport_type_for_modality(
        candidate["modality"]
    )

    return {
        "workoutName": (
            f"{candidate['prescription']} {candidate['modality']} - {candidate['workout_date']}"
        ),
        "sportType": sport_type,
        "estimatedDurationInSecs": sum(
            step["endConditionValue"] for step in steps
        ),
        "workoutSegments": [
            {
                "segmentOrder": 1,
                "sportType": sport_type,
                "workoutSteps": steps,
            }
        ],
    }

File: lib/services/garmin_workout_builder.dart
import 'fitr_workout_classifier.dart';

typedef GarminJson = Map<String, dynamic>;
typedef GarminGearTargetResolver =
    GarminGearTarget? Function(String prescription, String modality);

class GarminWorkoutBuilderException implements Exception {
  final String message;

  const GarminWorkoutBuilderException(this.message);

  @override
  String toString() => message;
}

class GarminGearTarget {
  final String metric;
  final String low;
  final String high;

  const GarminGearTarget({
    required this.metric,
    required this.low,
    required this.high,
  });
}

class GarminWorkoutBuilder {
  const GarminWorkoutBuilder();

  GarminJson buildCandidate(
    FitrWorkoutCandidate candidate, {
    required int age,
    GarminGearTargetResolver? gearTargetResolver,
  }) {
    if (age < 1 || age > 119) {
      throw const GarminWorkoutBuilderException(
        'Enter an age between 1 and 119.',
      );
    }

    return switch (candidate.type) {
      'MATT' => buildMattRowWorkout(),
      'Z2' => buildZone2Workout(candidate, age: age),
      'GEAR' => buildGearWorkout(
        candidate,
        gearTarget: gearTargetResolver?.call(
          candidate.prescription,
          candidate.modality,
        ),
      ),
      'MIXED_GEAR' => buildMixedGearWorkout(
        candidate,
        gearTargetResolver: gearTargetResolver,
      ),
      'POWER' => buildPowerWorkout(candidate),
      _ => throw GarminWorkoutBuilderException(
        'Unsupported FITR candidate type: ${candidate.type}',
      ),
    };
  }

  GarminJson timedStep(int order, int stepType, num seconds) {
    return {
      'type': 'ExecutableStepDTO',
      'stepOrder': order,
      'stepType': {'stepTypeId': stepType},
      'endCondition': {'conditionTypeId': 2, 'conditionTypeKey': 'time'},
      'endConditionValue': seconds,
      'targetType': {
        'workoutTargetTypeId': 1,
        'workoutTargetTypeKey': 'no.target',
      },
    };
  }

  GarminJson heartRateStep(
    int order,
    int stepType,
    num seconds, {
    num? low,
    num? high,
  }) {
    final step = timedStep(order, stepType, seconds);

    if (low != null || high != null) {
      step['targetType'] = {
        'workoutTargetTypeId': 4,
        'workoutTargetTypeKey': 'heart.rate.zone',
      };

      if (low != null) {
        step['targetValueOne'] = low;
      }

      if (high != null) {
        step['targetValueTwo'] = high;
      }
    }

    return step;
  }

  GarminJson sportTypeForModality(String modality) {
    final normalized = modality.toLowerCase();

    if (normalized.contains('run')) {
      return {'sportTypeId': 1, 'sportTypeKey': 'running'};
    }

    if (normalized.contains('c2 bike') || normalized == 'bike') {
      return {'sportTypeId': 2, 'sportTypeKey': 'cycling'};
    }

    return {'sportTypeId': 6, 'sportTypeKey': 'cardio_training'};
  }

  GarminJson buildMattRowWorkout() {
    const sportType = {'sportTypeId': 6, 'sportTypeKey': 'fitness_equipment'};

    return {
      'workoutName': 'M.A.T.T. Row Test',
      'estimatedDurationInSecs': 2400,
      'sportType': sportType,
      'workoutSegments': [
        {
          'segmentOrder': 1,
          'sportType': sportType,
          'workoutSteps': [
            timedStep(1, 3, 600.0),
            timedStep(2, 3, 1200.0),
            timedStep(3, 3, 600.0),
          ],
        },
      ],
    };
  }

  List<GarminJson> buildZone2Steps(
    FitrWorkoutCandidate candidate, {
    required int age,
  }) {
    final workSeconds = _requiredInt(candidate.classification, 'work_seconds');

    final hr160 = 160 - age;
    final hr165 = 165 - age;
    final hr170 = 170 - age;
    final hr180 = 180 - age;

    return [
      heartRateStep(1, 1, 300, low: 100, high: hr160),
      heartRateStep(2, 1, 300, low: hr160, high: hr165),
      heartRateStep(3, 1, 300, low: hr165, high: hr170),
      heartRateStep(4, 3, workSeconds, low: hr170, high: hr180),
      heartRateStep(5, 2, 300, low: hr165, high: hr170),
      heartRateStep(6, 2, 300, low: hr160, high: hr165),
      heartRateStep(7, 2, 300, low: 100, high: hr160),
    ];
  }

  GarminJson buildZone2Workout(
    FitrWorkoutCandidate candidate, {
    required int age,
  }) {
    final modality = candidate.modality;
    final normalized = modality.toLowerCase();

    final GarminJson sportType;

    if (normalized.contains('run')) {
      sportType = {'sportTypeId': 1, 'sportTypeKey': 'running'};
    } else if (normalized.contains('bike')) {
      sportType = {'sportTypeId': 2, 'sportTypeKey': 'cycling'};
    } else {
      sportType = {'sportTypeId': 6, 'sportTypeKey': 'cardio_training'};
    }

    return {
      'workoutName': 'Z2 $modality - ${candidate.date}',
      'sportType': sportType,
      'workoutSegments': [
        {
          'segmentOrder': 1,
          'sportType': sportType,
          'workoutSteps': buildZone2Steps(candidate, age: age),
        },
      ],
    };
  }

  double paceToSpeed(String pace, num distanceMeters) {
    final parts = pace.trim().split(':');

    if (parts.length != 2) {
      throw GarminWorkoutBuilderException('Invalid pace: $pace');
    }

    final minutes = int.tryParse(parts[0]);
    final seconds = double.tryParse(parts[1]);

    if (minutes == null || seconds == null) {
      throw GarminWorkoutBuilderException('Invalid pace: $pace');
    }

    final totalSeconds = minutes * 60 + seconds;

    if (totalSeconds <= 0) {
      throw GarminWorkoutBuilderException('Invalid pace: $pace');
    }

    return distanceMeters / totalSeconds;
  }

  double paceToMetersPerSecond(String pace) {
    return paceToSpeed(pace, 1609.344);
  }

  GarminJson applyRunPaceTarget(
    GarminJson step,
    String lowPace,
    String highPace,
  ) {
    final speeds = [
      paceToMetersPerSecond(lowPace),
      paceToMetersPerSecond(highPace),
    ];

    step['targetType'] = {
      'workoutTargetTypeId': 6,
      'workoutTargetTypeKey': 'pace.zone',
    };
    step['targetValueOne'] = speeds.reduce((a, b) => a > b ? a : b);
    step['targetValueTwo'] = speeds.reduce((a, b) => a < b ? a : b);

    return step;
  }

  String formatGearTarget(GarminGearTarget target) {
    const units = {
      'minPerMile': 'min/mile',
      'minPer500m': 'min/500m',
      'minPer1000m': 'min/1000m',
      'rpm': 'RPM',
    };

    final unit = units[target.metric];

    if (unit == null) {
      throw GarminWorkoutBuilderException(
        'Unsupported Gear target metric: ${target.metric}',
      );
    }

    final value = target.low == target.high
        ? target.low
        : '${target.low}-${target.high}';

    return 'Target $value $unit';
  }

  bool modalityUsesStructuredTarget(String modality) {
    final normalized = modality.toLowerCase();

    return (normalized.contains('run') ||
            normalized.contains('row') ||
            normalized.contains('bike')) &&
        !normalized.contains('echo');
  }

  GarminJson applyGearTarget(
    GarminJson step,
    GarminGearTarget target, {
    bool enforce = true,
  }) {
    step['description'] = formatGearTarget(target);

    if (!enforce) {
      return step;
    }

    if (target.metric == 'minPerMile') {
      return applyRunPaceTarget(step, target.low, target.high);
    }

    if (target.metric == 'minPer500m' || target.metric == 'minPer1000m') {
      final distance = target.metric == 'minPer500m' ? 500 : 1000;
      final speeds = [
        paceToSpeed(target.low, distance),
        paceToSpeed(target.high, distance),
      ];

      step['targetType'] = {
        'workoutTargetTypeId': 5,
        'workoutTargetTypeKey': 'speed.zone',
      };
      step['targetValueOne'] = speeds.reduce((a, b) => a < b ? a : b);
      step['targetValueTwo'] = speeds.reduce((a, b) => a > b ? a : b);

      return step;
    }

    throw GarminWorkoutBuilderException(
      'Structured target is not supported for metric: ${target.metric}',
    );
  }

  List<GarminJson> buildGearSteps(
    FitrWorkoutCandidate candidate, {
    GarminGearTarget? gearTarget,
  }) {
    final classification = candidate.classification;
    final rounds = _requiredInt(classification, 'rounds');
    final workSeconds = _requiredInt(classification, 'work_seconds');
    final restSeconds = _requiredInt(classification, 'rest_seconds');
    final paceLow = classification['pace_low']?.toString();
    final paceHigh = classification['pace_high']?.toString();
    final hasInlineRunPace =
        candidate.modality.toLowerCase() == 'run' &&
        paceLow != null &&
        paceLow.isNotEmpty &&
        paceHigh != null &&
        paceHigh.isNotEmpty;

    if (gearTarget == null && !hasInlineRunPace) {
      throw GarminWorkoutBuilderException(
        '${candidate.prescription} ${candidate.modality} requires a current '
        'athlete target. Restore or enter the target before building Garmin '
        'workouts.',
      );
    }

    final steps = <GarminJson>[];
    var order = 1;

    for (var round = 0; round < rounds; round++) {
      final workStep = timedStep(order, 3, workSeconds);

      if (gearTarget != null) {
        applyGearTarget(
          workStep,
          gearTarget,
          enforce: modalityUsesStructuredTarget(candidate.modality),
        );
      } else if (hasInlineRunPace) {
        applyRunPaceTarget(workStep, paceLow, paceHigh);
      }

      steps.add(workStep);
      order++;

      if (round < rounds - 1) {
        steps.add(timedStep(order, 4, restSeconds));
        order++;
      }
    }

    return steps;
  }

  GarminJson buildGearWorkout(
    FitrWorkoutCandidate candidate, {
    GarminGearTarget? gearTarget,
  }) {
    final steps = buildGearSteps(candidate, gearTarget: gearTarget);
    final sportType = sportTypeForModality(candidate.modality);

    return {
      'workoutName':
          '${candidate.prescription} ${candidate.modality} - ${candidate.date}',
      'sportType': sportType,
      'estimatedDurationInSecs': _totalDuration(steps),
      'workoutSegments': [
        {'segmentOrder': 1, 'sportType': sportType, 'workoutSteps': steps},
      ],
    };
  }

  GarminJson buildMixedGearWorkout(
    FitrWorkoutCandidate candidate, {
    GarminGearTargetResolver? gearTargetResolver,
  }) {
    final rawSteps = candidate.classification['steps'];

    if (rawSteps is! List) {
      throw const GarminWorkoutBuilderException(
        'Mixed Gear candidate is missing steps.',
      );
    }

    final workoutSteps = <GarminJson>[];

    for (var index = 0; index < rawSteps.length; index++) {
      final rawStep = rawSteps[index];

      if (rawStep is! Map) {
        throw const GarminWorkoutBuilderException(
          'Mixed Gear candidate contains an invalid step.',
        );
      }

      final candidateStep = Map<String, dynamic>.from(rawStep);
      final isWork = candidateStep['kind'] == 'work';
      final seconds = _requiredInt(candidateStep, 'seconds');
      final modality =
          candidateStep['modality']?.toString().trim().isNotEmpty == true
          ? candidateStep['modality'].toString()
          : 'Recovery';
      final rawPrescription = candidateStep['prescription']?.toString().trim();
      final hasStepPrescription =
          rawPrescription != null && rawPrescription.isNotEmpty;
      final prescription = hasStepPrescription
          ? rawPrescription
          : candidate.prescription;

      final step = timedStep(index + 1, isWork ? 3 : 4, seconds);

      final target = isWork
          ? gearTargetResolver?.call(prescription, modality)
          : null;

      if (isWork && target == null) {
        throw GarminWorkoutBuilderException(
          '$prescription $modality requires a current athlete '
          'target. Restore or enter the target before building Garmin '
          'workouts.',
        );
      }

      if (target != null) {
        applyGearTarget(
          step,
          target,
          enforce: modalityUsesStructuredTarget(modality),
        );
        final label = hasStepPrescription
            ? '$prescription $modality'
            : modality;
        step['description'] = '$label - ${formatGearTarget(target)}';
      } else {
        step['description'] = modality;
      }

      workoutSteps.add(step);
    }

    final GarminJson sportType = candidate.modality.contains('+')
        ? {'sportTypeId': 6, 'sportTypeKey': 'cardio_training'}
        : sportTypeForModality(candidate.modality);

    return {
      'workoutName':
          '${candidate.prescription} ${candidate.modality} - ${candidate.date}',
      'sportType': sportType,
      'estimatedDurationInSecs': _totalDuration(workoutSteps),
      'workoutSegments': [
        {
          'segmentOrder': 1,
          'sportType': sportType,
          'workoutSteps': workoutSteps,
        },
      ],
    };
  }

  List<GarminJson> buildPowerSteps(FitrWorkoutCandidate candidate) {
    final classification = candidate.classification;
    final rounds = _requiredInt(classification, 'rounds');
    final workSeconds = _requiredInt(classification, 'work_seconds');
    final recoverySeconds = _requiredInt(classification, 'recovery_seconds');

    final steps = <GarminJson>[];
    var order = 1;

    for (var round = 0; round < rounds; round++) {
      steps.add(timedStep(order, 3, workSeconds));
      order++;

      if (round < rounds - 1 && recoverySeconds > 0) {
        steps.add(timedStep(order, 4, recoverySeconds));
        order++;
      }
    }

    return steps;
  }

  GarminJson buildPowerWorkout(FitrWorkoutCandidate candidate) {
    final steps = buildPowerSteps(candidate);
    final sportType = sportTypeForModality(candidate.modality);

    return {
      'workoutName':
          '${candidate.prescription} ${candidate.modality} - ${candidate.date}',
      'sportType': sportType,
      'estimatedDurationInSecs': _totalDuration(steps),
      'workoutSegments': [
        {'segmentOrder': 1, 'sportType': sportType, 'workoutSteps': steps},
      ],
    };
  }

  int _requiredInt(Map<dynamic, dynamic> values, String key) {
    final value = values[key];

    if (value is int) {
      return value;
    }

    if (value is num) {
      return value.toInt();
    }

    final parsed = int.tryParse(value?.toString() ?? '');

    if (parsed == null) {
      throw GarminWorkoutBuilderException(
        'Candidate is missing a valid $key value.',
      );
    }

    return parsed;
  }

  num _totalDuration(List<GarminJson> steps) {
    return steps.fold<num>(
      0,
      (total, step) => total + (step['endConditionValue'] as num),
    );
  }
}
